File: evaluator.py
#!/usr/bin/env python3
"""
Independent second implementation of the IES_Policy billing evaluator, in a
different language from src/evaluator.js. Its whole purpose is to prove the
point Tariff Intelligence exists to prove: "multiple billing systems running
the same Policy Pack produce identical results" (IES v0.4 Strategy Document,
Section A1.6). It intentionally re-implements the same documented method
(see src/evaluator.js header) from scratch rather than porting the JS file
line-by-line, so agreement is evidence of a shared, unambiguous spec reading
-- not of copy-paste.

Usage:
    python evaluator.py <signed-policy.json> <usage-intervals.json>
Prints a JSON bill to stdout, in the same shape as evaluator.js's return value,
for test/run.js to diff against the Node result.
"""
import json
import re
import sys
from datetime import datetime, timezone


def round2(n):
    return round(n + 1e-9, 2)


def parse_iso_time_to_seconds(iso_time):
    m = re.match(r"^T?(\d{2}):(\d{2}):(\d{2})Z?$", iso_time)
    if not m:
        raise ValueError(f"Unrecognized interval.start format: {iso_time}")
    hh, mm, ss = (int(x) for x in m.groups())
    return hh * 3600 + mm * 60 + ss


def parse_iso_duration_to_seconds(iso_duration):
    m = re.match(r"^PT(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?$", iso_duration)
    if not m:
        raise ValueError(f"Unrecognized duration format: {iso_duration}")
    h, mi, s = (int(x) if x else 0 for x in m.groups())
    return h * 3600 + mi * 60 + s


def to_seconds_of_day(iso_timestamp):
    ts = iso_timestamp.replace("Z", "+00:00")
    d = datetime.fromisoformat(ts).astimezone(timezone.utc)
    return d.hour * 3600 + d.minute * 60 + d.second


def is_in_daily_window(iso_timestamp, interval):
    t = to_seconds_of_day(iso_timestamp)
    start = parse_iso_time_to_seconds(interval["start"])
    duration_sec = parse_iso_duration_to_seconds(interval["duration"])
    end = start + duration_sec
    if end <= 86400:
        return start <= t < end
    # wraps past midnight
    return t >= start or t < end - 86400


def compute_bill(policy, usage_intervals):
    total_kwh = sum(iv["kWh"] for iv in usage_intervals)
    slabs = sorted(policy["energySlabs"], key=lambda s: s["start"])

    remaining = total_kwh
    base_charge = 0.0
    slab_breakdown = []
    trace = []
    for slab in slabs:
        span = float("inf") if slab["end"] is None else (slab["end"] - slab["start"] + 1)
        kwh_in_slab = max(0.0, min(remaining, span))
        if kwh_in_slab <= 0:
            continue
        charge = kwh_in_slab * slab["price"]
        base_charge += charge
        remaining -= kwh_in_slab
        slab_breakdown.append({"slabId": slab["id"], "kWh": kwh_in_slab, "rate": slab["price"], "charge": round2(charge)})
        trace.append(f"Slab {slab['id']}: {kwh_in_slab} kWh x Rs{slab['price']} = Rs{round2(charge)}")
        if remaining <= 0:
            break

    blended_rate = (base_charge / total_kwh) if total_kwh > 0 else 0.0
    trace.append(f"Total {total_kwh} kWh, base energy charge Rs{round2(base_charge)}, blended rate Rs{round2(blended_rate)}/kWh")

    surcharge_breakdown = []
    adjustment_total = 0.0
    for s in policy.get("surchargeTariffs", []):
        kwh_in_window = sum(iv["kWh"] for iv in usage_intervals if is_in_daily_window(iv["startTime"], s["interval"]))
        if kwh_in_window <= 0:
            continue
        if s["unit"] == "INR_PER_KWH":
            amount = kwh_in_window * s["value"]
        else:
            amount = kwh_in_window * blended_rate * (s["value"] / 100)
        adjustment_total += amount
        surcharge_breakdown.append({"id": s["id"], "kWh": kwh_in_window, "unit": s["unit"], "value": s["value"], "amount": round2(amount)})
        unit_label = "%" if s["unit"] == "PERCENT" else " INR/kWh"
        trace.append(f"Surcharge {s['id']}: {kwh_in_window} kWh in window [{s['interval']['start']}, +{s['interval']['duration']}) x {s['value']}{unit_label} = Rs{round2(amount)}")

    total = round2(base_charge + adjustment_total)
    trace.append(f"Total bill = Rs{round2(base_charge)} + Rs{round2(adjustment_total)} = Rs{total}")

    return {
        "policyID": policy["policyID"],
        "policyRef": policy["id"],
        "totalKwh": total_kwh,
        "baseCharge": round2(base_charge),
        "slabBreakdown": slab_breakdown,
        "surchargeBreakdown": surcharge_breakdown,
        "total": total,
        "trace": trace,
    }


if __name__ == "__main__":
    envelope_path, usage_path = sys.argv[1], sys.argv[2]
    with open(envelope_path, encoding="utf-8") as f:
        envelope = json.load(f)
    policy = envelope["policy"] if "policy" in envelope else envelope
    with open(usage_path, encoding="utf-8") as f:
        usage_intervals = json.load(f)
    print(json.dumps(compute_bill(policy, usage_intervals), indent=2))

File: src/evaluator.js
// The "small evaluator" the docs describe: "look up the applicable slab, look
// up the matching time-of-day surcharge, apply." Consumes a verified IES_Policy
// object plus a billing period's interval usage and produces a bill + a full
// audit trace (the "bill-calculation trace" derived view named in the docs'
// Annexure D).
//
// BILLING METHOD (documented here because IES_Policy itself does not specify
// one — see ../proposed-extensions.md, Open Item #4):
//   1. Telescopic energy charge: total period kWh is allocated progressively
//      across energySlabs (first slab's span, then next, etc; a null `end`
//      means "unbounded top slab").
//   2. Time-of-day surchargeTariffs adjust the blended per-kWh rate for
//      whatever energy falls inside their recurring daily window. `interval`
//      windows that cross midnight (start + duration wraps past 24:00) are
//      handled explicitly.
//   3. PERCENT adjustments apply to the blended base rate; INR_PER_KWH
//      adjustments are a flat per-kWh add/subtract, independent of the base
//      rate.
export function computeBill(policy, usageIntervals) {
  const trace = { policyID: policy.policyID, policyRef: policy.id, steps: [] };

  // --- Step 1: telescopic slab allocation over total period kWh ---
  const totalKwh = usageIntervals.reduce((sum, iv) => sum + iv.kWh, 0);
  const slabs = [...policy.energySlabs].sort((a, b) => a.start - b.start);

  let remaining = totalKwh;
  let baseCharge = 0;
  const slabBreakdown = [];
  for (const slab of slabs) {
    const span = slab.end === null ? Infinity : slab.end - slab.start + 1;
    const kwhInSlab = Math.max(0, Math.min(remaining, span));
    if (kwhInSlab <= 0) continue;
    const charge = kwhInSlab * slab.price;
    baseCharge += charge;
    remaining -= kwhInSlab;
    slabBreakdown.push({ slabId: slab.id, kWh: kwhInSlab, rate: slab.price, charge: round2(charge) });
    trace.steps.push(`Slab ${slab.id}: ${kwhInSlab} kWh x Rs${slab.price} = Rs${round2(charge)}`);
    if (remaining <= 0) break;
  }
  const blendedRate = totalKwh > 0 ? baseCharge / totalKwh : 0;
  trace.steps.push(`Total ${totalKwh} kWh, base energy charge Rs${round2(baseCharge)}, blended rate Rs${round2(blendedRate)}/kWh`);

  // --- Step 2: time-of-day surcharge/discount adjustments ---
  const surchargeBreakdown = [];
  let adjustmentTotal = 0;
  for (const s of policy.surchargeTariffs ?? []) {
    const kwhInWindow = usageIntervals
      .filter((iv) => isInDailyWindow(iv.startTime, s.interval))
      .reduce((sum, iv) => sum + iv.kWh, 0);
    if (kwhInWindow <= 0) continue;

    const amount =
      s.unit === "INR_PER_KWH"
        ? kwhInWindow * s.value
        : kwhInWindow * blendedRate * (s.value / 100);

    adjustmentTotal += amount;
    surchargeBreakdown.push({ id: s.id, kWh: kwhInWindow, unit: s.unit, value: s.value, amount: round2(amount) });
    trace.steps.push(
      `Surcharge ${s.id}: ${kwhInWindow} kWh in window [${s.interval.start}, +${s.interval.duration}) x ${s.value}${s.unit === "PERCENT" ? "%" : " INR/kWh"} = Rs${round2(amount)}`
    );
  }

  const total = round2(baseCharge + adjustmentTotal);
  trace.steps.push(`Total bill = Rs${round2(baseCharge)} + Rs${round2(adjustmentTotal)} = Rs${total}`);

  return { policyID: policy.policyID, policyRef: policy.id, totalKwh, baseCharge: round2(baseCharge), slabBreakdown, surchargeBreakdown, total, trace: trace.steps };
}

// Does an ISO timestamp's time-of-day fall inside a recurring daily window,
// handling windows that wrap past midnight (e.g. 23:00 + 6h -> 05:00 next day)?
function isInDailyWindow(isoTimestamp, interval) {
  const t = toSecondsOfDay(isoTimestamp);
  const start = parseIsoTimeToSeconds(interval.start);
  const durationSec = parseIsoDurationToSeconds(interval.duration);
  const end = start + durationSec; // may exceed 86400 -> wraps past midnight

  if (end <= 86400) {
    return t >= start && t < end;
  }
  // Wraps past midnight: window is [start,86400) U [0, end-86400)
  return t >= start || t < end - 86400;
}

function toSecondsOfDay(isoTimestamp) {
  const d = new Date(isoTimestamp);
  return d.getUTCHours() * 3600 + d.getUTCMinutes() * 60 + d.getUTCSeconds();
}

// Parses "T18:00:00Z" (as used in the real devkit example) -> seconds since midnight.
function parseIsoTimeToSeconds(isoTime) {
  const m = /^T?(\d{2}):(\d{2}):(\d{2})Z?$/.exec(isoTime);
  if (!m) throw new Error(`Unrecognized interval.start format: ${isoTime}`);
  const [, hh, mm, ss] = m;
  return Number(hh) * 3600 + Number(mm) * 60 + Number(ss);
}

// Parses a simple ISO 8601 duration like "PT4H", "PT6H", "PT90M" -> seconds.
function parseIsoDurationToSeconds(isoDuration) {
  const m = /^PT(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?$/.exec(isoDuration);
  if (!m) throw new Error(`Unrecognized duration format: ${isoDuration}`);
  const [, h, min, s] = m;
  return (Number(h ?? 0) * 3600) + (Number(min ?? 0) * 60) + Number(s ?? 0);
}

function round2(n) {
  return Math.round(n * 100) / 100;
}
